src/category_theory/sites/cover_preserving.lean

```diff
@@ -0,0 +1,162 @@
+open category_theory.presieve.family_of_elements
+open category_theory.presieve
+open category_theory.limits
+
+namespace category_theory
+variables {C : Type u₁} [category.{v₁} C] {D : Type u₂} [category.{v₂} D]
+variables {A : Type u₃} [category.{v₃} A]
+variables (J : grothendieck_topology C) (K : grothendieck_topology D)
+variables {L : grothendieck_topology A}
+
+/--
+A functor `G : (C, J) ⥤ (D, K)` between sites is *cover-preserving*
+if for all covering sieves `R` in `C`, `R.pushforward_functor G` is a covering sieve in `D`.
+-/
+@[nolint has_inhabited_instance]
+structure cover_preserving (G : C ⥤ D) : Prop :=
+(cover_preserve : ∀ {U : C} {S : sieve U} (hS : S ∈ J U), S.functor_pushforward G ∈ K (G.obj U))
+
+/-- The identity functor on a site is cover-preserving. -/
+lemma id_cover_preserving : cover_preserving J J (𝟭 _) := ⟨λ U S hS, by simpa using hS⟩
+
+variables (J) (K)
+
+/-- The composition of two cover-preserving functors is cover-preserving. -/
+lemma cover_preserving.comp {G} (hG : cover_preserving J K G) {v} (hv : cover_preserving K L v) :
+  cover_preserving J L (G ⋙ v) := ⟨λ U S hS,
+begin
+  rw sieve.functor_pushforward_comp,
+  exact hv.cover_preserve (hG.cover_preserve hS)
+end⟩
+
+/--
+A functor `G : (C, J) ⥤ (D, K)` between sites is called compatible preserving if for each
+compatible family of elements at `C` and valued in `G.op ⋙ ℱ`, and each commuting diagram
+`f₁ ≫ G.map g₁ = f₂ ≫ G.map g₂`, `x g₁` and `x g₂` coincides when restricted via `fᵢ`.
+This is actually stronger than merely preserving compatible families because of the definition of
+`functor_pushforward` used.
+-/
+@[nolint has_inhabited_instance]
+structure compatible_preserving (K : grothendieck_topology D) (G : C ⥤ D) : Prop :=
+(compatible :
+  ∀ (ℱ : SheafOfTypes.{w} K) {Z} {T : presieve Z}
+    {x : family_of_elements (G.op ⋙ ℱ.val) T} (h : x.compatible)
+    {Y₁ Y₂} {X} (f₁ : X ⟶ G.obj Y₁) (f₂ : X ⟶ G.obj Y₂) {g₁ : Y₁ ⟶ Z} {g₂ : Y₂ ⟶ Z}
+    (hg₁ : T g₁) (hg₂ : T g₂) (eq : f₁ ≫ G.map g₁ = f₂ ≫ G.map g₂),
+      ℱ.val.map f₁.op (x g₁ hg₁) = ℱ.val.map f₂.op (x g₂ hg₂))
+
+variables {J K} {G : C ⥤ D} (hG : compatible_preserving.{w} K G) (ℱ : SheafOfTypes.{w} K) {Z : C}
+variables {T : presieve Z} {x : family_of_elements (G.op ⋙ ℱ.val) T} (h : x.compatible)
+
+include h hG
+
+/-- `compatible_preserving` functors indeed preserve compatible families. -/
+lemma presieve.family_of_elements.compatible.functor_pushforward :
+    (x.functor_pushforward G).compatible :=
+begin
+  rintros Z₁ Z₂ W g₁ g₂ f₁' f₂' H₁ H₂ eq,
+  unfold family_of_elements.functor_pushforward,
+  rcases get_functor_pushforward_structure H₁ with ⟨X₁, f₁, h₁, hf₁, rfl⟩,
+  rcases get_functor_pushforward_structure H₂ with ⟨X₂, f₂, h₂, hf₂, rfl⟩,
+  suffices : ℱ.val.map (g₁ ≫ h₁).op (x f₁ hf₁) = ℱ.val.map (g₂ ≫ h₂).op (x f₂ hf₂),
+    simpa using this,
+  apply hG.compatible ℱ h _ _ hf₁ hf₂,
+  simpa using eq
+end
+
+lemma compatible_preserving.apply_map {Y : C} {f: Y ⟶ Z} (hf) :
```

Comment:
I was gonna complain that you didn't write down the type of `hf`, but I see the merged version in #9691 has the type ...

Reply:
You could also see a comment asking me to add the type of `hf` over there. 😄